### DISCUSSION TITLE
LDM Notes for November 13th, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8591] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-11-13.md

## Agenda

- Extensions

**Comment by qrli** [comment]
```cs
    extension<T> (T) where T : class?
    {
        [param:NotNull] public void ThrowIfNull() => ArgumentNullException.ThrowIfNull(this);
    }
```
The `[param:NotNull]` does not feel like modifying `this` at all. I'd expect something like `[this:NotNull]`, which would be obvious even to people who read such syntax for the 1st time.

**Comment by qrli** [comment]
> Mixing underlying types within an extension declaration

How is
```cs
public static class CSharpExtensions
{
    extension
    {
        public bool (SyntaxToken).IsKind(SyntaxKind kind) => ...;
        public bool (SyntaxTrivia).IsKind(SyntaxKind kind) => ...;
        [param: NotNullWhen(true)]
        public bool (SyntaxNode?).IsKind(SyntaxKind kind) => ...;
        public bool (SyntaxNodeOrToken).IsKind(SyntaxKind kind) => ...;
        public bool (SyntaxNode).ContainsDirective(SyntaxKind kind) => ...;
    }
}
```
any better than the old
```cs
public static class CSharpExtensions
{
    public static bool IsKind(this SyntaxToken token, SyntaxKind kind) => ...;
    public static bool IsKind(this SyntaxTrivia trivia, SyntaxKind kind) => ...;
    public static bool IsKind([NotNullWhen(true)] this SyntaxNode? node, SyntaxKind kind) => ...;
    public static bool IsKind(this SyntaxNodeOrToken nodeOrToken, SyntaxKind kind) => ...;
    public static bool ContainsDirective(this SyntaxNode node, SyntaxKind kind) => ...;
}
```
if the old extension method syntax will still be around?

**Reply by CyrusNajmabadi** [reply]
> any better than the old

It can be consistently used for all instance and static extension *member* forms.  While the old one is only for instance extension *method*s.

**Reply by alrz** [reply]
> > any better than the old
> 
> It can be consistently used for all instance and static extension _member_ forms. While the old one is only for instance extension _method_s.

I believe the question is not about what it enables, rather, that it looks even more verbose than using `this` args and tbh a little bit forgiving on aesthetics.

 I think I'd really like to read arguments against type based approach since that's where this started with interface extentions in mind, but now it feels like it has been reduced to a sugar that almost doesn't read like C#, IMO.

**Reply by CyrusNajmabadi** [reply]
> rather, that it looks even more verbose than using this args 

That may be what is required to support all the different member forms. 

That said, it will *commonly* be *less verbose* when you have many extension members as you will not have to repeat the receiver N times (one per member). 

So, we believe it is a net win. It enables many new extension forms, and reduces overhead *per member*.  Even at a couple of members you're already ahead, and it improves the more you have. 

>  I think I'd really like to read arguments against type based approach

We ended up finding examples of virtually every type of extension out there in real packages. That meant the type approach ended up needing significant extensions itself to be a suitable replacement for existing extension methods. And those extensions were getting more and more ungainly. 

We had proposals on both. One approach felt like it solved all the use cases in a much more cohesive fashion. The other nailed a smaller set of cases, but felt really bad for a bunch of other important ones. We decided based on that.

**Reply by alrz** [reply]
> We decided based on that.

I understand there's tradeoffs to account for with either option, but I think this could be a style preference at the very least instead of only focusing on the migration story.

I'd imagine support for three forms (type, block, member) wouldn't be too much syntax since it's just a kind of modifier? I feel comfortable suggesting that because it's been considered to allow single-member extension as well as extension blocks. The former can address the indentation problem but there's no savings on the character count.

**Reply by HaloFour** [reply]
> I'd imagine support for three forms (type, block, member) wouldn't be too much syntax since it's just a kind of modifier?

I believe that was explicitly expressed in the meeting notes for the "unified" extensions proposal:, under "shorter forms": 
 https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-11-20.md

The syntax included is just illustrative of the concept.

**Reply by alrz** [reply]
>The syntax included is just illustrative of the concept.

The most interesting part in that is the instance members in a static class.

If we can get passed that, this would be the two forms next to each other:
```cs
static class E {
    extension(IEnumerable) {
        // instance members
    }
    extension<T>(IEnumerable<T>) {
        // instance members
    }
    // static members
}
```
```cs
extension(IEnumerable)
static partial class E {
  // instance members
}

extension<T>(IEnumerable<T>)
static partial class E {
  // instance members
}

static partial class E {
  // static members
}
```
With that, single-member form would actually save more when there's more regular members than extensions.
```cs
static partial class E {
   extension(int) 
   private bool IsZero => ..

   extension(string) 
   private bool IsEmpty => ..

  // static members
}
```
I assume explicit extension members cannot appear in another extension class though.

**Comment by jnm2** [comment]
I didn't say "entomology" to start with 😂

**Comment by markm77** [comment]
Just added a comment on extensions proposal [here](https://github.com/dotnet/csharplang/discussions/5496#discussioncomment-11298781).

**Comment by DavidArno** [comment]
I think there are some seriously flawed assumptions within this proposal:

> It is this author’s belief that programmers will resent being forced to separate extension declarations by underlying type...

As far as this programmer is concerned, the author is deeply mistaken. I always organise extension by the underlying type. Not doing so is messy and complicates the code in my view. Without clear empirical evidence, such guesses/beliefs risk delivering the wrong solution for the majority of devs.

Also, I do not understand why these extensions are being embedded in a static class. For me, I'd far prefer the example shown with the code:
```cs
public static class CSharpExtensions
{
    extension(SyntaxToken)
    {
        public bool IsKind(SyntaxKind kind) => ...;
```
were expressed as:
```cs
public extension(SyntaxToken)
{
    public bool IsKind(SyntaxKind kind) => ...;
}
    
public extension(SyntaxTrivia)
{
    public bool IsKind(SyntaxKind kind) => ...;
}

public extension(SyntaxNode?)
{
    [param: NotNullWhen(true)]
    public bool IsKind(SyntaxKind kind) => ...;
}

public extension(SyntaxNodeOrToken)
{
    public bool IsKind(SyntaxKind kind) => ...;
}

public extension(SyntaxNode)
{
    public bool ContainsDirective(SyntaxKind kind) => ...;
}
```
with the option to fall back to using `this` if I really need to:
```cs
public extension
{
    public bool IsKind(this SyntaxToken token, SyntaxKind kind) => ...;
    public bool IsKind(this SyntaxTrivia trivia, SyntaxKind kind) => ...;
    public bool IsKind([NotNullWhen(true)] this SyntaxNode? node, SyntaxKind kind) => ...;
    public bool IsKind(this SyntaxNodeOrToken nodeOrToken, SyntaxKind kind) => ...;
    public bool ContainsDirective(this SyntaxNode node, SyntaxKind kind) => ...;
}
```

The removal of all those noisy `static` keywords being motivation to rewrite existing extension methods into this form.

**Reply by CyrusNajmabadi** [reply]
> As far as this programmer is concerned, the author is deeply mistaken

The statement is not about all programmers. Just that these programmers do exist. This is a true statement as we've heard from people that this is a concern for them. As such, we're seeking inclusive solutions that work for the different sets of users. 

> Also, I do not understand why these extensions are being embedded in a static class.

Several reasons. Compat, for one. Allowing mixing of extensions and non extensions for another (this is super super super common). Providing a means for disambiguation. 

There are several reasons why having a named type is important, and why continuing to use static classes makes sense given how extension methods already work.

**Reply by DavidArno** [reply]
That is hugely disappointing. For me, the single most compelling reason for enhancing extension methods is to remove the need to explicitly create a static class for that extension. Often, I want an extension method that I only then use within one class. But it can't currently be declared locally and has to go in a separate static class. So I was anticipating something like this as a very basic requirement of this feature:
```cs
class SomeType<T>
{
    void Foo(List<T> list) => list.Transform();
    
    extension(List<T>)
    {
        void Transform() => ...;
    }
}
```
Without support for that, this feature just becomes some syntax to support extension properties etc. This seems super non-compelling to me as I don't think I've ever wanted a extension property to solve a problem, let alone an extension event handler.

Yet again this seems to be a "we're seeking inclusive solutions, but only for a given value of inclusive". Unsurprisingly, I guess, that "given value of inclusive" excludes me completely. Shame.

**Reply by CyrusNajmabadi** [reply]
> So I was anticipating something like this as a very basic requirement 

What you are considering to be requirements, we are considering as simple enhancements of the core proposal. We have many enhancements we think with be nice to have. But they have to be built on a solid foundation that meets all functional requirements first. Areas with less functional requirements can have increased ease of use features added to them.

**Reply by CyrusNajmabadi** [reply]
> excludes me completely.

How so? Nothing about your scenario is excluded. Indeed, many discussions have talked about taking extensions and making them available at virtually any sensible scope (including *within* a type or even a member). 

This is exactly what we mean by inclusive. We want a design that can scale and transform to all those needs. But we won't have such a design if we can't ensure that core functionality needs are met. 

Your need is syntactic sugar.  That is one of the easiest needs we can layer on a cohesive foundation. We've been designing entirely with that in mind.

**Reply by CyrusNajmabadi** [reply]
> This seems super non-compelling to me as I don't think I've ever wanted a extension property to solve a problem, 

Inclusivity means understanding that there are different users to you. Including massive libraries that have been asking for this (and other extension capabilities we are adding) for decades. You may not benefit from that, and have instead only cared about scoping an extension *inwards*, but many more groups and projects have cared about the core things we are designing. This is an inclusive design as those large groups can get what they want, while we can also do syntactic sugar to get you what you want.

**Reply by TahirAhmadov** [reply]
@DavidArno the gist of it is, they want to make the new extensions feature in a way that allows all existing extension methods (100%, without exception) be "transferable" to the new syntax and maintain binary compatibility.

1. Not only I don't really consider it to be that high of a priority - how often are you going to replace a DLL in your output with a massive set of changes, like being compiled against a whole new language version and probably new .NET version - without rebuilding your own code?
2. But also, in order to fulfill that "objective", they decided to ensure that the new extensions have to be emitted in a way which mimics existing extension methods. (Instead of relying on leaving existing extension methods as redirect stubs temporarily.)

And this approach opened a whole new can of worms. With redirect stubs, the user can see that the signature stayed the same, and binary compat is maintained.
But with new syntax, it's not apparent whatsoever. Therefore, the language needs to guarantee that. That's why a `static class` is necessary - because it's a way for the user to know how new extensions are emitted. (At least, it answers one part of that question. Other things, like generics and receiver parameter, you'll have to deduce yourself.)

Going forward, this means: we'll have existing extension methods (because we can't realistically deprecate these from the language); new extension members; and new extension "groups" within the `static class`.

> Allowing mixing of extensions and non extensions for another (this is super super super common). 

This is easily achieved in the syntactically "typed" form as well. Static classes aren't needed for this.

> Providing a means for disambiguation.

Ditto.

**Reply by TahirAhmadov** [reply]
Needless to say, what was originally envisioned as a straightforward, elegant feature has now become a little bit of a monstrosity.

**Reply by HaloFour** [reply]
The gist of it is that the language team is exploring the space given what is known about the existing ecosystem and design guidelines around extensions today.  No specific syntax decisions have been made, nor have extension types been ruled out.  Rather the syntax being thrown around are examples of ways that the concerns can be addressed, for the purposes of conversation.  The end result could be very different as the exact syntactic forms are the last part of the conversation.

Yes, the team is considering what it looks like to support migrating from existing extension methods.  Even if the team ultimately decides not to support it, it's their job to consider what it would look like.

**Reply by CyrusNajmabadi** [reply]
> With redirect stubs, the user can see that the signature stayed the same, and binary compat is maintained.

The last time this was looked at, no one could show that behavioral compat was maintained.  Has that changed?  

Note: binary comapt and source compat are not sufficient.  You need to ensure you get the same results when called.

**Reply by CyrusNajmabadi** [reply]
> without rebuilding your own code?

We have customers on diametrically opposite sides here.  You may be discussing how you do development, but we have to accept hte reality about how *broad swaths* of our development community do things as well.

**Reply by CyrusNajmabadi** [reply]
> But with new syntax, it's not apparent whatsoever. Therefore, the language needs to guarantee that

You have it incorrect.  By doing it in the language itself, we can *guarantee* things.  Having stubs, with different semantics on each side of the stub has no guarantees around compat, and you cannot 'visually' confirm anything.

**Reply by CyrusNajmabadi** [reply]
> This is easily achieved in the syntactically "typed" form as well.

Can you give an example of this?

**Reply by CyrusNajmabadi** [reply]
> Needless to say, what was originally envisioned as a straightforward,

There was nothing straightforward about the original release.  Both on the design and implementation side we ran into *enormous* problems.  That's why we were trying to prototype it out.  It was to see how viable it was.  We didn't guess at the unviability of it.  It was proven out by several compiler devs going down that path and determining just how problematic things were.  This was also confirmed by runtime team which came back with several hard blockers on the prior design that made is super undesirable to go that route.

**Reply by CyrusNajmabadi** [reply]
> Ditto.

What is your approach for disambiguation given David's preferred syntax?

**Reply by CyrusNajmabadi** [reply]
> be "transferable" to the new syntax and maintain binary compatibility.

This is not fully accurate.  Binary, source, and behavioral compat are *all* important to portions of the C# ecosystem.  We cannot blithely handwave that away just because we have small groups that are ok with breakage in any of these areas.

**Comment by MgSam** [comment]
Maybe I missed it, but in the new proposal how to do you statically call the new extension methods? Given they will be generated inside some form of unspeakable class. Being able to do call them statically is critical for disambiguation scenarios, no?

**Reply by CyrusNajmabadi** [reply]
It's an open question. But one we feel we have many possible solutions for. For example, we could raise existing syntax like so: `((MyEtensions)val).IsEmpty`. Or we could make new syntax.  We don't feel like this is a critical thing to solve currently.

**Reply by HaloFour** [reply]
It would be nice if there was a relatively simple way for downlevel (or non C#) languages to call extensions.  Keeping it simple also means that C# doesn't need to invent a syntax for that purpose.  IMO, that should be part of the conversation of lowering the extensions, especially since that represents a public API.
